external-crates/move/crates/move-compiler/src/diagnostics/mod.rs

```diff
@@ -28,19 +37,20 @@ use std::{
     iter::FromIterator,
     ops::Range,
     path::PathBuf,
+    sync::RwLock,
 };
 
 //**************************************************************************************************
 // Types
 //**************************************************************************************************
 
-#[derive(PartialEq, Eq, Clone, Debug, Hash)]
-#[must_use]
-pub struct Diagnostic {
-    info: DiagnosticInfo,
-    primary_label: (Loc, String),
-    secondary_labels: Vec<(Loc, String)>,
-    notes: Vec<String>,
+#[derive(Clone, Debug)]
+pub struct DiagnosticReporter<'env> {
+    flags: &'env Flags,
+    known_filter_names: &'env BTreeMap<DiagnosticsID, (FilterPrefix, FilterName)>,
+    diags: &'env RwLock<Diagnostics>,
+    ide_information: &'env RwLock<IDEInfo>,
+    warning_filters_scope: WarningFiltersScope,
```

Comment:
Apologies for the re-ordering here messing with the diff.

Reply:
I grabbed individual fields from `CompilationEnv` instead of just having a `&'env CompilationEnv` because I hope it will help ensure a better separation of abstractions for two reasons:
1. It should hopefully prevent feature creep that might occur if there was a direct reference to the env.
2. I hope this forces us to correctly construct a `DiagnosticReporter`. If `new` just took `&'env CompilationEnv`, it would very easy to create them incorrectly.

Reply:
That makes sense to me, and it also frees us from needing the whole environment just for reporting. Maybe we can even get some reuse this way.